### PR TITLE
657 typescript migration

### LINKED ISSUE #657: Support typescript development bundle
Add `npm run dev`

DoD:
- development goes through a bundle
- use map
- code in the map must be typescript
- bind watcher
- use rollup
- support npm package
- support umd and esm modules **(if possible)** with `npm run build`

FYI @tbo47 @yeluoyunfei @pavletto @maxbebop

### PR DESCRIPTION
**Pull request by Zemledelec**

### CODE REVIEW

File: src/og/Events.ts [hunk 1]
```diff
@@ -1,53 +1,66 @@
...
-     * @param {Array.<string>} [eventNames] - Event names that could be dispatched.
-     * @param {*} [sender]
+     * Registered event names.
+     * @protected
+     * @type {Array.<string>}
+     */
+    protected _eventNames: T;
+
+
+    protected _sender: any;
```
Comment (pavletto) [inline src/og/Events.ts]: I agree with sender any - it's could be any object from system
But disagree about `callback: any`! It's function, whith can get array of any parameters and return void or boolean
Disagree about _stampCache - it's Record<string, callback> 
And `as any` looks strange

propose decline any by eslint rools - it's a bad practice. Where we defenetly need any can off this rool by comment

Comment (pavletto) [reply src/og/Events.ts]: Also I think we need to declarate somehow created events in this class. this.event[event_name] shouldn't be `any`

File: src/og/input/KeyboardHandler.ts [hunk 2]
```diff
@@ -1,25 +1,41 @@
...
-        let _pressedKeysCallbacks = {};
-        let _unpressedKeysCallbacks = {};
-        let _charkeysCallbacks = {};
+    private _instance: any;
+
+    protected _removeCallback: any;
+    protected _getStamp: any;
+    protected _stamp: any;
+
+    public getcurrentlyPressedKeys: any;
```
Comment (pavletto) [inline src/og/input/KeyboardHandler.ts]: here I wanna cry( 
Try to descrybe all object instances as Record<some, some>

File: src/og/input/MouseHandler.ts [hunk 3]
```diff
@@ -2,24 +2,26 @@
...
         this._htmlObject = htmlObject;
     }
 
-    setEvent(event, sender, callback) {
+    public setEvent(event: string, sender: any, callback: Function) {
         switch (event) {
             case "mousewheel":
-                this._htmlObject.addEventListener('mousewheel', function (evt) {
-                    var delta = evt.deltaY || evt.detail || evt.wheelDelta;
+                this._htmlObject.addEventListener('mousewheel', function (evt: any) {
```
Comment (pavletto) [inline src/og/input/MouseHandler.ts]: any here is MouseEvent

File: src/og/input/MouseHandler.ts [hunk 4]
```diff
@@ -2,24 +2,26 @@
...
 
-    constructor(htmlObject) {
+    protected _htmlObject: HTMLElement;
+
+    constructor(htmlObject: HTMLElement) {
         this._htmlObject = htmlObject;
     }
 
-    setEvent(event, sender, callback) {
+    public setEvent(event: string, sender: any, callback: Function) {
```
Comment (pavletto) [inline src/og/input/MouseHandler.ts]: event here is <T extends Event> (lib.es event)

Comment (pavletto) [reply src/og/input/MouseHandler.ts]: or maybe MouseEvent

File: src/og/Events.ts (hunk 1 above)
Comment (Zemledelec) [reply src/og/Events.ts]: You could help! :)

File: src/og/input/KeyboardHandler.ts (hunk 2 above)
Comment (Zemledelec) [reply src/og/input/KeyboardHandler.ts]: No need to be so emotional. You could cry in another place:)

File: src/og/Clock.ts [hunk 5]
```diff
@@ -13,39 +27,42 @@ import { Events } from "./Events.js";
...
-        }
-        return this._counter;
-    }
+    static __counter__: number;
 
-    static set _staticCounter(n) {
-        this._counter = n;
-    }
+    public __handler: Handler | null = null;
+    public events: EventsHandler<ClockEventsList>;
```
Comment (tbo47) [inline src/og/Clock.ts]: the default visibility of a variable in typescript is `public` so there is no need to add it.

File: src/og/Events.ts [hunk 6]
```diff
@@ -112,17 +127,17 @@ class Events {
...
-    off(name, callback) {
+    public off(name: string, callback: any) {
         if (callback) {
-            var st = this._getStamp(name, this.__id, callback._openglobus_id);
+            let st = this._getStamp(name, this.__id, callback._openglobus_id);
             if (callback._openglobus_id && this._stampCache[st]) {
-                var h = this[name].handlers;
-                var i = h.length;
-                var indexToRemove = -1;
+                let h = (this as any)[name].handlers;
```
Comment (tbo47) [inline src/og/Events.ts]: you can replace `let` by `const`

File: src/og/Events.ts [hunk 7]
```diff
@@ -163,23 +178,21 @@ class Events {
...
 
     /**
      * Removes all events.
      * @public
      */
-    clear() {
+    public clear() {
         for (let i = 0; i < this._eventNames.length; i++) {
-            var e = this[this._eventNames[i]];
+            let e = (this as any)[this._eventNames[i]];
```
Comment (tbo47) [inline src/og/Events.ts]: `let` can be replaced by `const`

**Comment by tbo47** [comment]
I reviewed and tested it. It works great. Openglobus is going to be better with typescript. Thanks for doing this!

File: src/og/input/MouseHandler.ts (hunk 4 above)
Comment (tbo47) [reply src/og/input/MouseHandler.ts]: If we are not sure, we should just use `any`. We can refine this in a further PR. Just a suggestion.

**Comment by Zemledelec** [comment]
@pavletto, @tbo47 review, please :) I'm going to merge this.

P.S. No need to review 500 files, just the last few commits would be enough. I've checked it in our internal projects, and it's good.

**Pull request merged**